crates/relayer/src/chain/ckb4ibc/monitor.rs

```diff
@@ -0,0 +1,357 @@
+use std::str::FromStr;
+use std::sync::Arc;
+
+use ckb_ics_axon::handler::{IbcPacket, PacketStatus};
+use ckb_ics_axon::object::State as CkbState;
+use ckb_ics_axon::ChannelArgs;
+use ckb_jsonrpc_types::{Status, TransactionView};
+use ckb_sdk::rpc::ckb_indexer::SearchKey;
+use ckb_sdk::rpc::ckb_light_client::ScriptType;
+use ckb_types::packed::Script;
+use ckb_types::prelude::{Builder, Entity, Pack};
+use ckb_types::H256;
+use crossbeam_channel::Receiver;
+use ibc_relayer_types::core::ics02_client::height::Height;
+use ibc_relayer_types::core::ics03_connection::events::{
+    Attributes, OpenInit as ConnectionOpenInit, OpenTry as ConnectionOpenTry,
+};
+use ibc_relayer_types::core::ics04_channel::channel::State;
+use ibc_relayer_types::core::ics04_channel::events::{
+    OpenInit as ChannelOpenInit, OpenTry as ChannelOpenTry, ReceivePacket, SendPacket,
+};
+use ibc_relayer_types::core::ics04_channel::packet::Packet;
+use ibc_relayer_types::core::ics24_host::identifier::{ClientId, ConnectionId};
+use ibc_relayer_types::events::IbcEvent;
+use tokio::runtime::Runtime as TokioRuntime;
+
+use crate::chain::ckb::prelude::CkbReader;
+use crate::chain::ckb::rpc_client::RpcClient;
+use crate::chain::ckb4ibc::extractor::{
+    extract_channel_end_from_tx, extract_ibc_connections_from_tx, extract_ibc_packet_from_tx,
+};
+use crate::chain::tracking::TrackingId;
+use crate::config::ckb4ibc::ChainConfig;
+use crate::event::bus::EventBus;
+use crate::event::monitor::{Error, EventBatch, MonitorCmd, Next, Result, TxMonitorCmd};
+use crate::event::IbcEventWithHeight;
+
+// todo add cell emitter here
+pub struct Ckb4IbcEventMonitor {
+    rt: Arc<TokioRuntime>,
+    rpc_client: Arc<RpcClient>,
+    rx_cmd: Receiver<MonitorCmd>,
+    event_bus: EventBus<Arc<Result<EventBatch>>>,
+    config: ChainConfig,
+}
+
+impl Ckb4IbcEventMonitor {
+    pub fn new(
+        rt: Arc<TokioRuntime>,
+        rpc_client: Arc<RpcClient>,
+        config: ChainConfig,
+    ) -> (Self, TxMonitorCmd) {
+        let (tx_cmd, rx_cmd) = crossbeam_channel::unbounded();
+        let monitor = Ckb4IbcEventMonitor {
+            rt,
+            rpc_client,
+            rx_cmd,
+            event_bus: EventBus::default(),
+            config,
+        };
+        (monitor, TxMonitorCmd::new(tx_cmd))
+    }
+
+    pub fn run(mut self) {
+        let rt = self.rt.clone();
+        loop {
+            let result = rt.block_on(self.run_once());
+            match result {
+                Next::Continue => continue,
+                Next::Abort => break,
+            }
+        }
+    }
+    async fn run_once(&mut self) -> Next {
+        if let Ok(cmd) = self.rx_cmd.try_recv() {
+            match cmd {
+                MonitorCmd::Shutdown => return Next::Abort,
+                MonitorCmd::Subscribe(tx) => tx.send(self.event_bus.subscribe()).unwrap(),
+            }
+        }
```

Comment:
this code block should repeat before running `self.process_bach(result)`, because this is a try receive which can miss the outer subscribe request and directly enter in the dispatch process with no subscriber in there